### ISSUE TITLE
Automatic backup / Reminder for backuping database

### ISSUE THREAD

**Issue by Cotta-R** [cgeo/cgeo#6315]
I recently had (unplanned) to reinstall my smartphone from scratch. I tried to use the backup file of the database, but I noticed, it was quite old.

So I suggest to have a reminder every month (or maybe after a configurable time) to perform a new backup.

**Comment by GodfatherThe** [comment]
Maybe just "auto-backup" every week or so?

**Comment by SammysHP** [comment]
Or like most other backups: make a note in your calendar. It makes more sense to implement the backup API. IIRC there's already an issue for that.

**Comment by Lineflyer** [comment]
My opinion:
1. Auto backup
Theroretically a good thing, but has a lot ob obstacles (we need to check free space, we need to replace a backup which maybe a user made intentionally at a certain point and wants to keep it, what if we autobackup just in the moment something unwanted happend (e.g. user accidentally deleted a list), it would block or slow down the normal operation while executing, etc.)
Therefore I would be quite some effort to implement it in a clean way, not worth it IMHO.

2. Reminder:
Why not (if the user has the option to dismiss it forever if he does not want it). That should be an easy thing to do on startup as we have the date of the last backup and the current date.
It could be a first easy step before implementing a backup to some cloud service or something similar.

**Comment by Lineflyer** [comment]
As mentioned in https://github.com/cgeo/cgeo/issues/12479#issuecomment-1006419648

I would like to reuse this issue to go a step further as the original approach:

Quite often we have users on support who performed some action like deletion of caches, lists, offline logs and not using any of the available undo actions or only several days later recognize, that they made a mistake.
In such cases I ask users to check their backups, most of the time it turns out they did not have any.

Therefore just doing cyclic automatic backups is IMHO a convenience and safety feature to be added.
The easiest approach would be to just trigger it weekly in the background, which would enable (given the default of 5 stored backups) a backup time of the last 5 weeks.

Something more advanced would be to do it after a relevant change in database size/content is detected or any other clever approach. But thats probably nothing which will ever fit all needs.

**Comment by moving-bits** [comment]
Doing an automatic backup every x weeks will eventually overwrite the oldest backup - which might not be wanted by some users. So maybe better stay with automatic reminder?

We could do it similar to the way we have for map & routing data updates: Remind the user every x days to create a backup, offering a button to directly start the backup (and another to postpone it). Similar to the other functions, user can set up a frequency (with frequency = 0 meaning "no reminders"), default would be frequency = 7 (days).

**Comment by Lineflyer** [comment]
If you newly install c:geo doing an automatic backup will not hurt users as nothing will be overwritten.

How about introducing a switch for auto backup which is initially set to active if no backup is present at the time of upgrade.

**Comment by moving-bits** [comment]
You're talking about "auto backup instead of auto reminder" or both?

If the first, we could use
- a checkbox "auto backups on/off", and
- a slider to determine the frequency (or would we leave out this, and set frequency hard-coded to 7 days?)

If the letter, we could use
- a switch "auto backup vs. auto reminder", and
- a slider to determine the frequency (with value 0=off)

The second would give us more flexibility while staying in line UI-wise with the other auto-reminder functions.

Default would be:
- if no backup available the first time: activate "auto backup" with "frequency=7",
- otherwise: activate "auto reminder" with "frequency=7".

What do you think?

**Comment by eddiemuc** [comment]
My 2ct:
**If** we introduce automatic backup, those should be strictly separated from any manual Backup the user might do. Preferably e.g. in a separate folder named "auto" or whatever. An automatic backup should never overwrite or lead to the deletion of a manual Backup.

**Comment by Lineflyer** [comment]
This proposal is even better. Just introduce an automatic backup into a seperate folder. That would not conflict with any manual backups.

I would think a reminder is just annoying...lets do it automatic.

**Comment by moving-bits** [comment]
Would we want to have a single automatic backup, which gets overwritte every time a new backup is created, or do we want to create a stack of (e. g.: 5) auto backups? (IMHO having a small stack would make sense.)

And I guess those auto backups should be available in a public folder to be accessible by the user, and to not get deleted automatically on uninstalling c:geo. Folder could be `auto` beneath our already existing `backup` folder.

**Comment by eddiemuc** [comment]
> Would we want to have a single automatic backup, which gets overwritte every time a new backup is created, or do we want to create a stack of (e. g.: 5) auto backups? (IMHO having a small stack would make sense.)

Imho this makes sense. 5 Backups once a week might be a sensible default. but this should be somehow configurable (number of Backups and frequency) because users might not all have so much memory on their phone.


> And I guess those auto backups should be available in a public folder to be accessible by the user, and to not get deleted automatically on uninstalling c:geo. Folder could be `auto` beneath our already existing `backup` folder.

Yes, I guess it would not make much sense to delete those on uninstall

**Comment by eddiemuc** [comment]
> 5 Backups once a week might be a sensible default

Sorry, I meant one Backup per week and keep the last 5 Backups

**Comment by eddiemuc** [comment]
One Thing though: I think we can't do Backups in the background since the db needs to be free of connections for this - correct? So I assume once c:geo wants to do an automatic backup it has to ask the user anyway (he can't use c:geo for a short amount of time then)

**Comment by moving-bits** [comment]
> but this should be somehow configurable (number of Backups and frequency) because users might not all have so much memory on their phone.

yes, should be configurable (slider for frequency with 0=off)

> one Backup per week and keep the last 5 Backups

shall we reuse the existing setting (for manual backups), or introduce a separate one?

**Comment by moving-bits** [comment]
One additional thought why asking the user before starting a backup might be a good idea anyway: 
Imagine you're "out in the wild", wanting to cache, starting c:geo - and it starts a (probably lenghty*) backup session first... Being able to postpone this for a day may be good thing. (* depending on system performance and amount of data)

**Comment by eddiemuc** [comment]
> shall we reuse the existing setting (for manual backups), or introduce a separate one?

IMHO a separate one

**Comment by fm-sys** [comment]
> One additional thought why asking the user before starting a backup might be a good idea anyway:
> Imagine you're "out in the wild", wanting to cache, starting c:geo - and it starts a (probably lenghty*) backup session first...

True, but keep in mind many users use cgeo only once a week. Do you really want to annoy them every time they use cgeo?

Another solution would be to do the backup automatically via a background service as soon as the user closes cgeo. Not sure if this is technically possible though...
